New serverless pattern - DynamoDB Streams to Step Functions using Event bridge pipes (#1922)
aws-samples/serverless-patterns: eventbridge-pipes-ddbstream-sfn/README.md

# DynamoDB Stream to Step Functions with EventBridge Pipes

> This pattern shows how to use EventBridge Pipes to launch a StepFunctions state machine with a message coming from DynamoDB Stream

Reviewer: ```suggestion
This pattern shows how to use Amazon EventBridge Pipes to launch a AWS Step Functions state machine with a message coming from an Amazon DynamoDB stream
```

* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
* [AWS Serverless Application Model](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) (AWS SAM) installed

## Deployment Instructions

1. Create a new directory, navigate to that directory in a terminal and clone the GitHub repository:
    ``` 
    git clone https://github.com/aws-samples/serverless-patterns
    ```
1. Change directory to the pattern directory:
    ```
    cd eventbridge-pipes-ddbstream-sfn
    ```
1. From the command line, use AWS SAM to deploy the AWS resources for the pattern as specified in the template.yml file:
    ```
    sam deploy --guided
    ```
1. During the prompts:
    * Enter a stack name
    * Enter the desired AWS Region
    * Allow SAM CLI to create IAM roles with the required permissions.

    Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in future to use these defaults.

1. Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs which are used for testing.

## How it works

> Previously, whenever we needed DynamoDB Streams to send messages to Step Functions based on changes made to your DynamoDB table, we had to implement a Lambda function in between DynamoDB and Step Functions because DynamoDB streams only supported Lambda as its destination.

Reviewer: DynamoDB Streams also supported Kinesis Data Streams. I suggest to reword it as "because DynamoDB Streams did not support Step Functions as a target directly"

> But at re:invent 2022, AWS launched a new feature for AWS EventBridge known as EventBridge Pipes.

Reviewer: ```suggestion
But at re:Invent 2022, AWS launched a new feature for Amazon EventBridge known as EventBridge Pipes.
```

Now, we can directly integrate DynamoDB streams with AWS Step Functions without the need for having an additional Lambda function.

## Testing

> 1. Stream logs from StepFunctions LogGroup

Reviewer: ```suggestion
1. Stream logs from the Step Functions log group
```

```
sam logs --cw-log-group <LogGroup Name> --tail
```

> 2. The Pipe is configured to filter on the NationalTeam value of "Argentina"

Reviewer: ```suggestion
2. The EventBridge Pipe is configured to filter on the NationalTeam value of "Argentina"
```

> In another terminal Add an item to the DynamoDB stream which does match the filter.

Reviewer: ```suggestion
In another terminal, add an item to the DynamoDB stream which does match the filter.
```

```bash
aws dynamodb put-item \
    --table-name WorldCupTable \
    --item PlayerName={S="Lionel Messi"},Nationality={S="Argentina"},GoalsScored={S="1"}
```

> Step Function should be invoked and you should see the logs for the new execution.

Reviewer: ```suggestion
The Step Functions workflow should be invoked and you should see the logs for the new execution.
```

Now add an item to the DynamoDB stream which doesn't match the filter.
```bash
aws dynamodb put-item \
    --table-name WorldCupTable \
    --item PlayerName={S="Sergy Gnabry"},Nationality={S="Germany"},GoalsScored={S="1"}
```

> Step Function should not be invoked and you should not see any new logs.

Reviewer: ```suggestion
No Step Function workflow should be invoked and you should not see any new logs.
```
